langchain/src/retrievers/pinecone-hybrid-search.ts

```diff
@@ -0,0 +1,148 @@
+import type {
+  SparseValues,
+  VectorOperationsApi,
+} from "@pinecone-database/pinecone/dist/pinecone-generated-ts-fetch";
+import { BertTokenizer } from "bert-tokenizer";
+
+import { PineconeLibArgs, PineconeMetadata } from "vectorstores/pinecone.js";
+import { Embeddings } from "../embeddings/base.js";
+import { Document } from "../document.js";
+import { BaseRetriever } from "../schema/index.js";
+
+export interface PineconeHybridSearchParams extends PineconeLibArgs {
+  tokenizer: BertTokenizer;
+  topK: number;
+  alpha: number;
+}
+
+export class PineconeHybridSearchRetriever extends BaseRetriever {
+  embeddings: Embeddings;
+
+  pineconeIndex: VectorOperationsApi;
+
+  tokenizer: BertTokenizer;
+
+  topK = 4;
+
+  alpha = 0.5;
+
+  textKey: string;
+
+  namespace?: string;
+
+  filter?: PineconeMetadata;
+
+  constructor(embeddings: Embeddings, args: PineconeHybridSearchParams) {
+    super();
+    this.embeddings = embeddings;
+    this.pineconeIndex = args.pineconeIndex;
+    this.tokenizer = args.tokenizer;
+    this.topK = args.topK;
+    this.alpha = args.alpha;
+    this.namespace = args.namespace;
+    this.textKey = args.textKey ?? "text";
+    this.filter = args.filter;
+  }
+
+  buildDict(inputBatch: number[][]): SparseValues[] {
+    return inputBatch.map((tokenIds) => {
+      const tokenCounts = tokenIds.reduce((acc, tokenId) => {
+        acc[tokenId] = (acc[tokenId] || 0) + 1;
+        return acc;
+      }, {} as Record<number, number>);
+
+      return {
+        indices: Object.keys(tokenCounts).map(Number),
+        values: Object.values(tokenCounts),
+      };
+    });
+  }
+
+  async addTexts(texts: string[], metadatas: object[] | object): Promise<void> {
+    const batchSize = 32;
+
+    for (let i = 0; i < texts.length; i += batchSize) {
+      const iEnd = Math.min(i + batchSize, texts.length);
+      const contextBatch = texts.slice(i, iEnd);
+      const ids = Array.from({ length: iEnd - i }, (_, idx) =>
+        (i + idx).toString()
+      );
+
+      const denseEmbeds = await this.embeddings.embedDocuments(contextBatch);
+      const tokenIdsBatch = contextBatch.map((text) =>
+        this.tokenizer.tokenize(text)
+      );
+      const sparseEmbeds = this.buildDict(tokenIdsBatch);
+
+      const vectors = ids.map((id, idx) => {
+        const metadata = Array.isArray(metadatas)
+          ? metadatas[i + idx]
+          : metadatas;
+
+        return {
+          id,
+          sparse_values: sparseEmbeds[idx],
+          values: denseEmbeds[idx],
+          metadata: { [this.textKey]: contextBatch[idx], ...metadata },
+        };
+      });
+
+      await this.pineconeIndex.upsert({
+        upsertRequest: { vectors, namespace: this.namespace },
+      });
+    }
+  }
+
+  hybridScale(
+    dense: number[],
+    sparse: SparseValues,
+    alpha: number
+  ): [number[], SparseValues] {
+    if (alpha < 0 || alpha > 1) {
+      throw new Error("Alpha must be between 0 and 1");
```

Comment:
_nit_: consider validating `this.alpha` in the constructor. 

_nit_: do we need to pass `this.alpha` as a parameter? Why not access it directly?

Reply:
_nit_: consider validating `this.alpha` in the constructor. 
_nit_: do we need to pass `this.alpha` as a parameter? Why not access it directly?